File: 백준/Gold/1976. 여행 가자/여행 가자.py
n = int(input())
m = int(input())
l = [0] * (n + 1)

def union(a, b):
    a2 = find(a)
    b2 = find(b)
    if a2 != b2:
        l[b2] = a2

def find(a):
    if a == l[a]:
        return a
    else:
        l[a] = find(l[a])
        return l[a]

for i in range(n + 1):
    l[i] = i

for i in range(n):
    t = list(map(int, input().split()))
    for k in range(len(t)):
        if t[k] == 1:
            union(i + 1, k + 1)

route = list(map(int, input().split()))
r = find(route[0])
isPossible = True
for i in range(1, len(route)):
    if find(route[i]) != r:
        isPossible = False
        break

if isPossible:
    print('YES')
else:
    print('NO')

File: 백준/Gold/1976. 여행 가자/여행 가자.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    public static int[] parents;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int N = Integer.parseInt(br.readLine());
        int M = Integer.parseInt(br.readLine());

        parents = new int[N + 1];

        for(int i = 1; i <= N; i++) {
            parents[i] = i;
        }

        for(int i = 1; i <= N; i++) {
            st = new StringTokenizer(br.readLine());

            for(int j = 1; j <= N; j++) {
                if(Integer.parseInt(st.nextToken()) == 1) {
                    union(i, j);
                }
            }
        }

        int[] route = new int[M];

        st = new StringTokenizer(br.readLine());
        for(int i = 0; i < M; i++) {
            route[i] = Integer.parseInt(st.nextToken());
        }

        boolean isPossible = true;

        for(int i = 0; i < M - 1; i++) {
            int start = route[i];
            int end = route[i + 1];

            if(find(start) != find(end)) {
                isPossible = false;
                break;
            }
        }

        if (isPossible) {
            System.out.println("YES");
        } else {
            System.out.println("NO");
        }
    }

    public static int find(int a) {
        if(a == parents[a]) {
            return a;
        }

        return find(parents[a]);
    }

    public static void union(int a, int b) {
        int pA = find(a);
        int pB = find(b);

        if(pA != pB) {
            parents[pB] = pA;
        }
    }
}